### PR TITLE
Baditsa/changing workshops status

### PR DESCRIPTION
**Pull request by valerabad**
Changing status for workshop entity.
What is done:
- new endpoint UpdateStatus
- workshop status field has been added to the filter
   - possible values for the filter: "Open", "Closed", "Open, Closed"
   - by default - all (filter isn't applied)
- notifications for applied parents and who liked the workshop
- creating an application is possible when the workshop status is open
- the only private workshops shall have the ability to close the workshop
- validation: workshop status field can be set up by "Open" or 1, "Closed" or 2 values

### CODE REVIEW

File: OutOfSchool/OutOfSchool.WebApi/Services/ApplicationService.cs [hunk 1]
```diff
@@ -400,6 +409,18 @@ public async Task<IEnumerable<string>> GetNotificationsRecipientIds(Notification
             return recipientIds.Distinct();
         }
 
+        private async Task<WorkshopStatus> CheckWorkshopStatus(Guid id)
+        {
+            var workshop = await workshopRepository.GetById(id).ConfigureAwait(false);
```
Comment (RBiyovskiy) [inline OutOfSchool/OutOfSchool.WebApi/Services/ApplicationService.cs:414]: The method does not match the name, in my opinion. It simply returns the status, not checks it.

File: OutOfSchool/OutOfSchool.WebApi.Tests/Services/ApplicationServiceTests.cs [hunk 2]
```diff
@@ -187,6 +190,71 @@ public void CreateApplication_WhenParametersAreNotValid_ShouldThrowArgumentExcep
...
+                async () => await service.Create(input).ConfigureAwait(false));
+        }
+
+        [Test]
+        public async Task CreateApplication_WhenStatusWorkshopIsOpen_ShouldReturnApplication()
+        {
+            // Arrange
+            var newApplication = new Application()
+            {
+                Id = new Guid("6d4caeae-f0c3-492e-99b0-c8c105693376"),
```
Comment (RBiyovskiy) [inline OutOfSchool/OutOfSchool.WebApi.Tests/Services/ApplicationServiceTests.cs:224]: Please, use predefined variable, instead than used the same code some times.

File: OutOfSchool/OutOfSchool.WebApi/Services/Database/WorkshopService.cs [hunk 3]
```diff
@@ -245,6 +251,35 @@ public async Task<WorkshopDTO> Update(WorkshopDTO dto)
             return mapper.Map<WorkshopDTO>(currentWorkshop);
         }
 
+        /// <inheritdoc/>
+        public async Task<WorkshopDTO> UpdateStatus(Guid id, WorkshopStatus status)
+        {
+            logger.LogInformation($"Updating Workshop status with Id = {id} started.");
+
+            var currentWorkshop = await workshopRepository.GetWithNavigations(id).ConfigureAwait(false);
```
Comment (i-rakityanskiy) [inline OutOfSchool/OutOfSchool.WebApi/Services/Database/WorkshopService.cs:259]: We do not need all of the workshop related information here. This call can be changed.
```suggestion
            var currentWorkshop = await workshopRepository.GetById(id).ConfigureAwait(false);
```

File: OutOfSchool/OutOfSchool.WebApi/Services/Database/WorkshopService.cs [hunk 4]
```diff
@@ -245,6 +251,35 @@ public async Task<WorkshopDTO> Update(WorkshopDTO dto)
             return mapper.Map<WorkshopDTO>(currentWorkshop);
         }
 
+        /// <inheritdoc/>
+        public async Task<WorkshopDTO> UpdateStatus(Guid id, WorkshopStatus status)
```
Comment (i-rakityanskiy) [inline OutOfSchool/OutOfSchool.WebApi/Services/Database/WorkshopService.cs:255]: Please add tests for this method.

File: OutOfSchool/OutOfSchool.WebApi.Tests/Controllers/WorkshopControllerTests.cs [hunk 5]
```diff
@@ -298,7 +301,24 @@ public async Task UpdateWorkshop_WhenModelIsValid_ShouldReturnOkObjectResult()
             Assert.That(result, Is.Not.Null);
             Assert.AreEqual(Ok, result.StatusCode);
         }
+        #endregion
+
+        #region UpdateWorkshopStatus
```
Comment (i-rakityanskiy) [inline OutOfSchool/OutOfSchool.WebApi.Tests/Controllers/WorkshopControllerTests.cs:306]: Please add more tests for different scenarios.

File: OutOfSchool/OutOfSchool.WebApi.Tests/Controllers/WorkshopControllerTests.cs [hunk 6]
```diff
@@ -298,7 +301,24 @@ public async Task UpdateWorkshop_WhenModelIsValid_ShouldReturnOkObjectResult()
...
+        [Test]
+        public async Task UpdateWorkshopStatus_WhenModelIsValid_ShouldReturnOkObjectResult()
```
Comment (i-rakityanskiy) [inline OutOfSchool/OutOfSchool.WebApi.Tests/Controllers/WorkshopControllerTests.cs:308]: Word `Workshop` is already present  in the class name. Test name could be simplified.
```suggestion
        public async Task UpdateStatus_WhenModelIsValid_ShouldReturnOkObjectResult()
```

File: OutOfSchool/OutOfSchool.WebApi/Services/ApplicationService.cs [hunk 7]
```diff
@@ -90,6 +91,14 @@ public class ApplicationService : IApplicationService, INotificationReciever
 
             ModelNullValidation(applicationDto);
 
+            var workshopStatus = await CheckWorkshopStatus(applicationDto.WorkshopId).ConfigureAwait(false);
+
+            if (workshopStatus == WorkshopStatus.Closed)
+            {
+                logger.LogInformation("Unable to create a new application for a workshop because workshop status is closed.");
+                throw new ArgumentException("Unable to create a new application for a workshop because workshop status is closed.");
+            }
```
Comment (i-rakityanskiy) [inline OutOfSchool/OutOfSchool.WebApi/Services/ApplicationService.cs:100]: `CheckWorkshopStatus` does not perform the status checking. You should probably move this piece of code inside the `CheckWorkshopStatus` method.

**Review by i-rakityanskiy — CHANGES_REQUESTED** [review]

File: OutOfSchool/OutOfSchool.WebApi/Services/WorkshopServicesCombiner.cs [hunk 8]
```diff
@@ -185,6 +213,18 @@ private bool IsFilterValid(WorkshopFilter filter)
             return filter != null && filter.MaxStartTime >= filter.MinStartTime
                                   && filter.MaxAge >= filter.MinAge
                                   && filter.MaxPrice >= filter.MinPrice;
+        } 
+        
+        public async Task<IEnumerable<string>> GetNotificationsRecipientIds(NotificationAction action, Dictionary<string, string> additionalData, Guid objectId)
+        {
+            var recipientIds = new List<string>();
+
+            var usersInFavoriteWorkshop = await favoriteService.GetAll().ConfigureAwait(false);
```
Comment (i-rakityanskiy) [inline OutOfSchool/OutOfSchool.WebApi/Services/WorkshopServicesCombiner.cs:222]: This call is too heavy for the DB. Please use filter here.

File: OutOfSchool/OutOfSchool.WebApi/Services/WorkshopServicesCombiner.cs [hunk 9]
```diff
@@ -185,6 +213,18 @@ private bool IsFilterValid(WorkshopFilter filter)
             return filter != null && filter.MaxStartTime >= filter.MinStartTime
                                   && filter.MaxAge >= filter.MinAge
                                   && filter.MaxPrice >= filter.MinPrice;
+        } 
+        
+        public async Task<IEnumerable<string>> GetNotificationsRecipientIds(NotificationAction action, Dictionary<string, string> additionalData, Guid objectId)
```
Comment (i-rakityanskiy) [inline OutOfSchool/OutOfSchool.WebApi/Services/WorkshopServicesCombiner.cs:218]: In the story mentioned that 
> all users that liked the workshop and applied see notifications about the change. 

I'm not sure what does _**applied users**_ mean and why we should notify them. It is better to ask BA.

**Comment by i-rakityanskiy** [comment]
Build of tests failed. Please take the latest changes from the `develop` branch.

File: OutOfSchool/OutOfSchool.WebApi/Common/WorkshopUpdateStatusException.cs [hunk 10]
```diff
@@ -0,0 +1,10 @@
+using System;
+
+namespace OutOfSchool.WebApi.Common
+{
+    public class WorkshopUpdateStatusException : Exception
+    {
+        public WorkshopUpdateStatusException(string message)
+            : base(message) { }
```
[the remark this answers is not in the document]
Comment (valerabad) [reply OutOfSchool/OutOfSchool.WebApi/Common/WorkshopUpdateStatusException.cs:8]: What predefined type can I use? I use these custom exceptions to handle 3 different logic cases

File: OutOfSchool/OutOfSchool.WebApi/Services/WorkshopServicesCombiner.cs (hunk 8 above)
Comment (valerabad) [reply OutOfSchool/OutOfSchool.WebApi/Services/WorkshopServicesCombiner.cs:222]: Please ask what filter should I use?

File: OutOfSchool/OutOfSchool.WebApi.Tests/Controllers/WorkshopControllerTests.cs [hunk 11]
```diff
@@ -332,6 +334,83 @@ public async Task UpdateWorkshop_WhenModelIsInvalid_ShouldReturnBadRequestObject
...
+        {
+            // Arrange
+            providerServiceMoq.Setup(x => x.GetByUserId(It.IsAny<string>(), It.IsAny<bool>())).ReturnsAsync(provider);
+
+            var updateRequest = new WorkshopStatusDto()
+            {
+                WorkshopId = workshop.Id,
+                Status = WorkshopStatus.Open,
+            };
+
```
Comment (RBiyovskiy) [inline OutOfSchool/OutOfSchool.WebApi.Tests/Controllers/WorkshopControllerTests.cs:349]: var updateRequest = new WorkshopStatusDto()
            {
                WorkshopId = workshop.Id,
                Status = WorkshopStatus.Open,
            };
            
            You use the same code some times. Will be better move it to the another method and add needed status as the method's parameter.

File: OutOfSchool/OutOfSchool.WebApi/Controllers/V1/WorkshopController.cs [hunk 12]
```diff
@@ -218,17 +220,54 @@ public async Task<IActionResult> Update(WorkshopDTO dto)
...
+        [HasPermission(Permissions.WorkshopEdit)]
+        [ProducesResponseType(StatusCodes.Status200OK, Type = typeof(WorkshopStatusDto))]
+        [ProducesResponseType(StatusCodes.Status400BadRequest)]
+        [ProducesResponseType(StatusCodes.Status401Unauthorized)]
+        [ProducesResponseType(StatusCodes.Status403Forbidden)]
+        [ProducesResponseType(StatusCodes.Status500InternalServerError)]
+        [HttpPut]
+        public async Task<IActionResult> UpdateStatus([FromBody] WorkshopStatusDto request)
+        {
+            var workshop = await combinedWorkshopService.GetById(request.WorkshopId).ConfigureAwait(false);
```
Comment (RBiyovskiy) [inline OutOfSchool/OutOfSchool.WebApi/Controllers/V1/WorkshopController.cs:250]: Please, validate parameter 'request' is non-null before using it.
Please review also other sonar issues.

Comment (i-rakityanskiy) [reply OutOfSchool/OutOfSchool.WebApi/Controllers/V1/WorkshopController.cs:255]: There is no need for such a request validation, this check is performed by the framework.

File: OutOfSchool/OutOfSchool.WebApi.Tests/Controllers/WorkshopControllerTests.cs [hunk 13]
```diff
@@ -647,5 +715,15 @@ private WorkshopDescriptionItemDto FakeWorkshopDescriptionItem()
                 Description = $"test description text sentence for id: {id.ToString()}",
             };
         }
+
+        private WorkshopStatusDto FakeWorkshopStatusDto(Guid workshopDtoId, WorkshopStatus workshopStatus)
```
Comment (i-rakityanskiy) [inline OutOfSchool/OutOfSchool.WebApi.Tests/Controllers/WorkshopControllerTests.cs:719]: It is good practice to use verbs in method names to describe what the method performs. I understand that `Fake` can be used a verb but in this case it does not reflect the methods purpose. Let's rename this method to something like `WithWorkshop` or `GetWorkshop`.

Comment (i-rakityanskiy) [inline OutOfSchool/OutOfSchool.WebApi.Tests/Controllers/WorkshopControllerTests.cs:719]: It is good practice to use verbs in method names to describe what the method performs. I understand that `Fake` can be used a verb but in this case it does not reflect the methods purpose. Let's rename this method to something like `WithWorkshopStatusDto` or `GetWorkshopStatusDto`.

File: OutOfSchool/OutOfSchool.WebApi/Services/WorkshopServicesCombiner.cs [hunk 14]
```diff
@@ -171,13 +207,27 @@ public async Task<List<WorkshopCard>> GetByProviderId(Guid id)
...
             await workshopService.GetWorkshopProviderOwnerIdAsync(workshopId).ConfigureAwait(false);
 
-        private List<WorkshopCard> DtoModelsToWorkshopCards(IEnumerable<WorkshopDTO> source)
+        public async Task<IEnumerable<string>> GetNotificationsRecipientIds(NotificationAction action, Dictionary<string, string> additionalData, Guid objectId)
         {
-            return source.Select(currentElement => currentElement.ToCardDto()).ToList();
+            var recipientIds = new List<string>();
+
+            Expression<Func<Favorite, bool>> filter = x => x.WorkshopId == objectId;
+            var favoriteWorkshopUsersIds = favoriteRepository.Get(where: filter).Select(x => x.UserId).ToList();
```
Comment (i-rakityanskiy) [inline OutOfSchool/OutOfSchool.WebApi/Services/WorkshopServicesCombiner.cs:219]: There is not need in the `filter` variable. Please change this to async call.
```suggestion
            var favoriteWorkshopUsersIds = await favoriteRepository.Get(where: x => x.WorkshopId == objectId)
                .Select(x => x.UserId)
                .ToListAsync()
                .ConfigureAwait(false);
```

File: OutOfSchool/OutOfSchool.WebApi/Services/WorkshopServicesCombiner.cs [hunk 15]
```diff
@@ -171,13 +207,27 @@ public async Task<List<WorkshopCard>> GetByProviderId(Guid id)
...
+
+            Expression<Func<Application, bool>> predicate =
+                x => x.Status != ApplicationStatus.Left &&
+                     x.WorkshopId == objectId;
+            var applications = await applicationRepository.GetByFilter(predicate).ConfigureAwait(false);
+            var appliedUsersIds = applications.Select(x => x.Parent.UserId);
```
Comment (i-rakityanskiy) [inline OutOfSchool/OutOfSchool.WebApi/Services/WorkshopServicesCombiner.cs:225]: Performance-wise it is better to use Select on IQueryable, i.e. before you make call to the DB
```suggestion
            var appliedUsersIds = await applicationRepository.Get(where: predicate)
                .Select(x => x.Parent.UserId)
                .ToListAsync()
                .ConfigureAwait(false);
```

**Review by i-rakityanskiy — CHANGES_REQUESTED** [review]

File: OutOfSchool/OutOfSchool.WebApi.Tests/Services/Database/WorkshopServiceTests.cs [hunk 16]
```diff
@@ -235,6 +238,77 @@ public void Update_WhenClassIdIsInvalid_ShouldThrowArgumentOutOfRangeException()
 
         #endregion
 
+        #region UpdateStatus
+        [Test]
+        public async Task UpdateStatus_FromOpenToClosed_WhenEntityIsValid_ShouldReturnUpdatedEntity()
+        {
+            // Arrange
+            var workshopStatusDtoMock = WithWorkshopsList().
+                FirstOrDefault(w => w.Status == WorkshopStatus.Open && w.ProviderOwnership == OwnershipType.Private);
```
Comment (i-rakityanskiy) [inline OutOfSchool/OutOfSchool.WebApi.Tests/Services/Database/WorkshopServiceTests.cs:247]: Period at the end of the line is a bit confusing. Please move it to the beginning of the next line. Do the same in the code below.

**Review by i-rakityanskiy — DISMISSED** [review]

**Review by i-rakityanskiy — APPROVED** [review]

**Pull request merged**